local-cluster/tests/local_cluster.rs

```diff
@@ -3762,6 +3786,222 @@ fn test_run_test_load_program_accounts_partition_root() {
+    let full_snapshot_archive_info = snapshot_utils::bank_to_full_snapshot_archive(
+        ledger_path,
+        &bank,
+        Some(SnapshotVersion::default()),
+        ledger_path,
+        ArchiveFormat::TarZstd,
+        1,
+        1,
+    )
+    .unwrap();
+    info!(
+        "Successfully created snapshot for slot {}, hash {}: {}",
+        bank.slot(),
+        bank.hash(),
+        full_snapshot_archive_info.path().display(),
+    );
+}
+
+fn do_test_hard_fork_with_or_without_gap_in_roots(with_gap: bool) {
+    solana_logger::setup_with_default(RUST_LOG_FILTER);
+
+    // First set up the cluster with 2 nodes
+    let slots_per_epoch = 2048;
+    let node_stakes = vec![60, 40];
+
+    let validator_keys = vec![
+        "28bN3xyvrP4E8LwEgtLjhnkb7cY4amQb6DrYAbAYjgRV4GAGgkVM2K7wnxnAS7WDneuavza7x21MiafLu1HkwQt4",
+        "2saHBBoTkLMmttmPQP8KfBkcCw45S5cwtV3wTdGCscRC8uxdgvHxpHiWXKx4LvJjNJtnNcbSv5NdheokFFqnNDt8",
+    ]
+    .iter()
+    .map(|s| (Arc::new(Keypair::from_base58_string(s)), true))
+    .take(node_stakes.len())
+    .collect::<Vec<_>>();
+    let validators = validator_keys
+        .iter()
+        .map(|(kp, _)| kp.pubkey())
+        .collect::<Vec<_>>();
+
+    let validator_a_pubkey = validators[0];
+    let validator_b_pubkey = validators[1];
+
+    let validator_config = ValidatorConfig {
+        snapshot_config: Some(LocalCluster::create_dummy_load_only_snapshot_config()),
+        no_hard_fork_blockstore_root_reconcilation_for_local_cluster_test: with_gap,
+        ..ValidatorConfig::default()
+    };
+    let mut config = ClusterConfig {
+        cluster_lamports: 100_000,
+        node_stakes: node_stakes.clone(),
+        validator_configs: make_identical_validator_configs(&validator_config, node_stakes.len()),
+        validator_keys: Some(validator_keys),
+        slots_per_epoch,
+        stakers_slot_offset: slots_per_epoch,
+        skip_warmup_slots: true,
+        ..ClusterConfig::default()
+    };
+    let cluster = std::sync::Arc::new(std::sync::Mutex::new(LocalCluster::new(
+        &mut config,
+        SocketAddrSpace::Unspecified,
+    )));
+
+    let val_a_ledger_path = cluster.lock().unwrap().ledger_path(&validator_a_pubkey);
+    let val_b_ledger_path = cluster.lock().unwrap().ledger_path(&validator_b_pubkey);
+
+    let min_last_vote = 45;
+    let min_root = 10;
+    loop {
+        sleep(Duration::from_millis(100));
+
+        if let Some((last_vote, _)) = last_vote_in_tower(&val_a_ledger_path, &validator_a_pubkey) {
+            if last_vote >= min_last_vote
+                && root_in_tower(&val_a_ledger_path, &validator_a_pubkey) > Some(min_root)
+            {
+                break;
+            }
+        }
+    }
+
+    let mut validator_a_info = cluster.lock().unwrap().exit_node(&validator_a_pubkey);
+    let mut validator_b_info = cluster.lock().unwrap().exit_node(&validator_b_pubkey);
+
+    let hard_fork_slot = min_last_vote - 5;
```

Comment:
maybe an assertion here after node exit that the `hard_fork_slot` is greater than the root_in_tower, to ensure no weird race condition happened after node exit

Reply:
done.